documents/extension/funding/service.go

```diff
@@ -175,23 +175,45 @@ func (s service) DeriveFromPayload(ctx context.Context, req *clientfunpb.Funding
 	var fd Data
 	fd.initFundingFromData(req.Data)
 
-	model, err := s.GetCurrentVersion(ctx, identifier)
+	fd.BorrowerId = req.Data.BorrowerId
+	fd.FunderId = req.Data.FunderId
+
+	funderID, err := identity.NewDIDFromString(req.Data.FunderId)
 	if err != nil {
-		apiLog.Error(err)
-		return nil, documents.ErrDocumentNotFound
+		return nil, err
 	}
-
-	ca, err := documents.FromClientCollaboratorAccess(req.ReadAccess, req.WriteAccess)
+	borrowerID, err := identity.NewDIDFromString(req.Data.BorrowerId)
 	if err != nil {
 		return nil, err
 	}
-
+	model, err := s.GetCurrentVersion(ctx, identifier)
+	if err != nil {
+		apiLog.Error(err)
+		return nil, documents.ErrDocumentNotFound
+	}
 	attributes, err := createAttributesList(model, fd)
 	if err != nil {
 		return nil, err
 	}
 
-	err = model.AddAttributes(ca, true, attributes...)
+	// check which Id needs to be added as a new collaborator to the document
+	var c []identity.DID
+	for _, id := range []identity.DID{funderID, borrowerID} {
+		collaborator, err := model.IsDIDCollaborator(id)
+		if err != nil {
+			return nil, err
+		}
+		if !collaborator {
+			c = append(c, id)
+		}
+	}
+	err = model.AddAttributes(
+		documents.CollaboratorsAccess{
+			ReadWriteCollaborators: c,
```

Comment:
I think we would lose new collaborators from `req.ReadAccess, req.WriteAccess`. It could contain new collaborators which are not funders or borrowers.

I think we should
` ca, err := documents.FromClientCollaboratorAccess(req.ReadAccess, req.WriteAccess)`

Afterwards, append your `c` to `ca.ReadWriteCollaborators`.

What do you think?

Reply:
We removed req.ReadAccess and req.WriteAccess since they serve the purpose of adding funder ID to the collaborator list.

> Afterwards, append your c to ca.ReadWriteCollaborators.
is the funder supoosed to get the write access ?